src/stream/src/executor/source/source_executor.rs

```diff
@@ -252,26 +255,8 @@ impl<S: StateStore> SourceExecutor<S> {
             self.actor_ctx.id.to_string(),
             core.source_id.to_string(),
         ]);
-        // fetch the newest offset, either it's in cache (before barrier)
-        // or in state table (just after barrier)
-        let target_state = if core.state_cache.is_empty() {
-            for ele in &mut *split_info {
-                if let Some(recover_state) = core
-                    .split_state_store
-                    .try_recover_from_state_store(ele)
-                    .await?
-                {
-                    *ele = recover_state;
-                }
-            }
-            split_info.to_owned()
-        } else {
-            core.state_cache
-                .values()
-                .map(|split_impl| split_impl.to_owned())
-                .collect_vec()
-        };
 
+        let target_state = core.latest_split_info.values().cloned().collect();
```

Comment:
I have some correctness concerns here. Seems you are resetting the process to the offset where the last successful barrier came, taking the time as T0. 
But later an error occurs at time T1 and requires rebuilding source **internally**. The logic here may lead to read data from T0 to T1 twice.

Reply:
I think the expected logic here is taking a union of both `state_cache` and `split_info` to make sure resetting every assigned split to its latest offset.

Reply:
> Seems you are resetting the process to the offset where the last successful barrier came

No. `latest_split_info` is also updated on every message chunk, so its offset is always up to date.

Reply:
This is indeed very confusing and also why I wanted to use just one.

![image.png](https://graphite-user-uploaded-assets-prod.s3.amazonaws.com/JgO8A8bpio3B4YlQByok/c65c751e-7445-4464-9abd-d66b59803637.png)